qmi/instruments/thorlabs/mpc320.py

```diff
@@ -0,0 +1,507 @@
+        """
+        if not self.MIN_VELOCITY_PERC <= vel <= self.MAX_VELOCITY_PERC:
+            raise QMI_InstrumentException(
+                f"Given relative velocity {vel} is outside the valid range \
+                    [{self.MIN_VELOCITY_PERC}%, {self.MAX_VELOCITY_PERC}%]"
+            )
+
+    def _validate_channel(self, channel_number: int) -> None:
+        """
+        Validate the channel number. The MPC320 has 3 channels.
+
+        Parameters:
+            channel_number: Channel number to validate.
+
+        Raises:
+            an instance of QMI_InstrumentException if the channel is not 1, 2 or 3
+        """
+
+        if channel_number not in [1, 2, 3]:
+            raise QMI_InstrumentException(
+                f"Given channel {channel_number} is not in the valid range \
+                    [{self.MIN_CHANNEL_NUMBER}, {self.MAX_CHANNEL_NUMBER}]"
+            )
+
+    @rpc_method
+    def open(self) -> None:
+        _logger.info("[%s] Opening connection to instrument", self._name)
+        self._check_is_closed()
+        self._transport.open()
+        super().open()
+
+    @rpc_method
+    def close(self) -> None:
+        _logger.info("[%s] Closing connection to instrument", self._name)
+        super().close()
+        self._transport.close()
+
+    @rpc_method
+    def get_idn(self) -> QMI_InstrumentIdentification:
+        """
+        Read instrument type and version and return QMI_InstrumentIdentification instance.
+
+        Returns:
+            an instance of QMI_InstrumentIdentification. The version refers to the firmware version.
+        """
+        _logger.info("[%s] Getting identification of instrument", self._name)
+        self._check_is_open()
+        # Send request message.
+        self._apt_protocol.write_param_command(AptMessageId.HW_REQ_INFO.value)
+        # Get response
+        resp = self._apt_protocol.ask(HW_GET_INFO)
+        return QMI_InstrumentIdentification("Thorlabs", resp.model_number, resp.serial_number, resp.firmware_version)
+
+    @rpc_method
+    def identify(self) -> None:
+        """
+        Identify device by flashing the front panel LEDs.
+        """
+        _logger.info("[%s] Identifying device", self._name)
+        self._check_is_open()
+        # Send message.
+        # For the MPC320 the channel number does not matter here. The device has one LED that flashes irrespective
+        # of the provided channel number.
+        self._apt_protocol.write_param_command(AptMessageId.MOD_IDENTIFY.value, 0x01)
+
+    @rpc_method
+    def enable_channels(self, channel_numbers: List[int]) -> None:
+        """
+        Enable the channel(s). Note that this method will disable any channel that is not provided as an argument. For
+        example, if you enable channel 1, then 2 and 3 will be disabled. If you have previously enabled a channel(s)
+        and fail to include it/them again in this call, that channel(s) will be disabled. For example, if you run the
```

Comment:
This must be the Thorlabs logic for this function? As I don't see anywhere a 'disable' list.

Reply:
yes, for this driver, you automatically disable channels that are not in the enable list. That's why for the `disable_all_channels()` command you will notice that I use the same message ID `MOD_SET_CHANENABLESTATE` but with an empty list. So I am enabling no channels, which automatically disables all of them